StenosisMeasurement3D/qSlicerStenosisMeasurement3DModuleWidget.cxx

```diff
@@ -355,56 +428,294 @@ void qSlicerStenosisMeasurement3DModuleWidget::onTubePointEndInteraction(vtkObje
     this->logic->UpdateBoundaryControlPointPosition(0, fiducialNode, shapeNode);
     this->logic->UpdateBoundaryControlPointPosition(1, fiducialNode, shapeNode);
   }
+  if (d->ParameterNode)
+  {
+    d->ParameterNode->SetInputFiducialNodeID(this->inputFiducialNode ? this->inputFiducialNode->GetID() : nullptr);
+  }
 }
 
 //-----------------------------------------------------------------------------
 void qSlicerStenosisMeasurement3DModuleWidget::onShapeNodeChanged(vtkMRMLNode * node)
 {
-  if (this->currentShapeNode == node)
+  Q_D(qSlicerStenosisMeasurement3DModuleWidget);
+  if (this->inputShapeNode == node)
   {
     return;
   }
-  if (this->currentShapeNode)
+  if (this->inputShapeNode)
   {
     // Disconnect the currently observed node.
-    this->currentShapeNode->RemoveObserver(this->tubeObservation);
+    this->inputShapeNode->RemoveObserver(this->tubeObservation);
   }
-  this->currentShapeNode = node;
-  if (currentShapeNode)
+  this->inputShapeNode = node;
+  if (inputShapeNode)
   {
     // Connect the current node.
-    this->currentShapeNode->AddObserver(vtkMRMLMarkupsNode::PointEndInteractionEvent, this->tubeObservation);
+    this->inputShapeNode->AddObserver(vtkMRMLMarkupsNode::PointEndInteractionEvent, this->tubeObservation);
   }
-  // Move control points to closest point on spline.
-  if (this->currentShapeNode && this->currentFiducialNode)
+  // Move control points to closest point on spline. See notes in ::onFiducialNodeChanged().
+  if (this->inputShapeNode && this->inputFiducialNode && d->ParameterNode)
   {
-    vtkMRMLMarkupsFiducialNode * fiducialNode = vtkMRMLMarkupsFiducialNode::SafeDownCast(this->currentFiducialNode);
-    vtkMRMLMarkupsShapeNode * shapeNode = vtkMRMLMarkupsShapeNode::SafeDownCast(this->currentShapeNode);
+    vtkMRMLMarkupsFiducialNode * fiducialNode = vtkMRMLMarkupsFiducialNode::SafeDownCast(d->ParameterNode->GetInputFiducialNode());
+    vtkMRMLMarkupsShapeNode * shapeNode = vtkMRMLMarkupsShapeNode::SafeDownCast(this->inputShapeNode);
     this->logic->UpdateBoundaryControlPointPosition(0, fiducialNode, shapeNode);
     this->logic->UpdateBoundaryControlPointPosition(1, fiducialNode, shapeNode);
   }
+  if (d->ParameterNode)
+  {
+    d->ParameterNode->SetInputShapeNodeID(this->inputShapeNode ? this->inputShapeNode->GetID() : nullptr);
+  }
 }
 
+//-----------------------------------------------------------------------------
+void qSlicerStenosisMeasurement3DModuleWidget::onSegmentationNodeChanged(vtkMRMLNode * node)
+{
+  Q_D(qSlicerStenosisMeasurement3DModuleWidget);
+  if (!this->inputSegmentationNode)
+  {
+    /*
+     * The segmentation selector is special.
+     * If we don't clear it explicitly, the last segment is selected
+     * on CloseScene/ReOpenSavedScene/SelectSameSegmentation.
+     */
+    d->inputSegmentSelector->setCurrentSegmentID(nullptr);
+  }
+  if (this->inputSegmentationNode == node)
+  {
+    return;
+  }
+  this->inputSegmentationNode = node;
```

Comment:
Do we really need this class member? Couldn't we just store this segmentation node in the parameter node?

Reply:
These similar class members are removed, the widget relies on the parameter node only, a single storage simplifies things.